### DISCUSSION TITLE
【腾讯犀牛鸟2026】Penguin-VL-2B 的 ncnn 移植与双平台部署

### DISCUSSION THREAD

**Discussion by LiamFan16-mikasa** [Tencent/ncnn#6840] [Show and tell]
# 【腾讯犀牛鸟2026】Penguin-VL-2B 的 ncnn 移植与双平台部署

> 代码仓库：https://github.com/LiamFan16-mikasa/Penguin-VL-ncnn
> 对应 issue：Tencent/ncnn #6788

## 任务目标

将 `tencent/Penguin-VL-2B`（视觉编码器 28 层 + Qwen3 系语言模型 28 层）从 PyTorch/Hugging Face 推理链路移植到 ncnn，交付一个纯 C++、运行时仅依赖 ncnn 与 stb_image 的多模态推理程序。程序读取一张图片和任意文本 prompt，在 C++ 内完成全部工作并输出文本：

1. 读取并预处理图片；
2. 用 Qwen tokenizer 编码任意文本 prompt；
3. 构造 Penguin-VL 多模态 chat template；
4. 用 ncnn vision encoder 把图片转换为 image embeddings；
5. 把 image embeddings 注入文本 token embedding；
6. 用并行 prefill 建立每一层的 KV cache；
7. 用 KV cache 逐 token decode，并把 token id 解码回文本。

## 完成情况

- vision encoder、projector 与 Qwen3 LLM 的 ncnn 模型转换（共拆为 114 张可独立验证的子图）；
- C++ 图片预处理、Qwen byte-level BPE tokenizer 与 chat template；
- 512-token 并行 prefill 与定长窗口 KV-cache decode；
- 权重逐图流式加载（低内存）与常驻两种运行模式；
- Linux（GCC）与 Windows（MinGW-w64）双平台 CMake 构建与 CPU 端到端运行；
- 在相同精度（fp32）下，C++ ncnn 与 PyTorch greedy decode 的 128-token 输出逐 id 完全一致。

## 结果展示

**测试输入**：一张 252×252 的方形图片；prompt 为 `描述这个图片`。

**运行命令**：

```
./penguinvl --image <图片> --prompt "描述这个图片" --model <模型目录> --max-new-tokens 128
```

**对齐结果**：截图上半部分为 C++ ncnn runtime 输出，下半部分为 PyTorch fp32 参考。两端使用相同图片、相同 prompt、greedy 解码、最多 128 个新 token，逐 id 输出完全一致（128/128），并在同一步生成 EOS。

![C++ ncnn 与 PyTorch fp32 逐 id 对齐（128/128）](assets/align_128.png)
<img width="635" height="434" alt="align_128" src="https://github.com/user-attachments/assets/46727197-7263-4668-a354-c6129bbbecb3" />
Windows（MinGW-w64）平台构建并运行，输出与 Linux 一致：
![Windows（MinGW-w64）构建运行](assets/windows_run.png)
<img width="628" height="129" alt="windows_run" src="https://github.com/user-attachments/assets/c3261326-4c5d-4b05-92c2-2cbb95b399bd" />

**逐级对齐数据**（均在 fp32 下测得）：视觉链路 28 层 transform 余弦相似度 ≥ 0.999985，端到端 projector 余弦相似度 0.9970；LLM prefill 全层余弦相似度 ≥ 0.99993，首 token 与参考一致；decode 逐 token 对齐通过；C++ 端到端 128/128 token 逐 id 一致。

## 解码与采样说明

当前采用 greedy argmax 解码，不启用随机采样。每一步经 final RMSNorm 与 lm_head 得到整个词表的 logits 后取概率最高的 token——只有确定性解码才能与 PyTorch 逐 id 对齐。回答质量取决于模型本身与采样策略，不属于移植正确性范畴；移植的验收标准是"相同输入下与参考逐 token 一致"，而非"回答是否准确"。

### 1. 建立 PyTorch 参考输出

移植前先运行原始 `tencent/Penguin-VL-2B`，dump 一整套 ground truth：输入 token ids、image token 数量与位置、pixel_values、vision 输出、projector 输出、融合后的 embeddings、逐层 hidden、prefill logits、128 个 greedy token 与最终文本。此后每完成一个 ncnn 模块，都与该参考逐级比对；排查顺序固定为 输入处理 → vision encoder → projector → LLM layer → KV cache → lm_head → token，不依据最终文本猜测错误来源。

一处必须注意：参考须以 **fp32** 生成。若参考用 bf16 而实现用 fp32，长程 greedy 会在某步两候选 logit 极接近时被精度差翻边，产生近义词替换的"分叉"，这并非移植错误（详见"实现难点 6"）。此外，官方 loader 在 CPU 上需传 `attn_implementation="sdpa"`，视觉侧的 `flash_attn_varlen_func` 需注入等价实现，`process_text` 会原地把 input_ids 的 image 槽位改写为 0，dump 前须先快照。

### 2. C++ 图片预处理

`image_processor` 直接读取图片，执行与 PyTorch processor 一致的处理：

```
RGB 图片
  -> bicubic resize（对齐 patch=14）
  -> rescale 1/255
  -> mean=0.5、std=0.5 归一化
  -> 切分为 14×14 patch
  -> 每个 RGB patch 展平为 3×14×14=588 维
  -> pixel_values shape = [N, 588]
```

两处细节决定成败：目标尺寸计算须用银行家舍入（round-half-to-even），与 Python 一致，相差一个像素则后续全错；bicubic 须复刻 Pillow 的定点实现（8bit 定点，precision=22），纯浮点公式会产生 ±1 灰阶差异。验收标准是与 HF processor 输出零灰阶差异。vision RoPE 亦在 C++ 生成，与图片数据一起送入 ncnn，无需 Python 中间文件。

### 3. 图片转 image embeddings

视觉部分拆为 30 张 ncnn 子图串联：

```
[N, 588] pixel_values
  -> vision embedding（patch conv 改写为等价 Linear）
  -> 28 个 vision transformer layer
  -> vision norm
  -> projector（mlp2x_gelu, 1024→2048）
  -> [N, 2048] image embeddings
```

image embeddings 不是 tokenizer 产生的整数 id，而是 N 个连续的 2048 维向量。projector 把视觉 hidden size 映射到语言模型 hidden size，使视觉向量可放入 LLM 的 embedding 序列。视觉链路全程 fp32（原因见"实现难点 4"）。

### 4. C++ 编码任意 prompt

`qwen_tokenizer` 加载 `vocab.json`、`merges.txt` 与 `tokenizer_config.json`，实现 Qwen byte-level BPE：

```
UTF-8 文本
  -> special token 最长优先匹配
  -> byte-level 映射
  -> BPE 贪心合并
  -> token ids
```

`prompt_builder` 按 Penguin-VL 的 chat template 组合 system/user/assistant 标记，并在图片位置插入 image token 占位符（id=151669）。验收标准是与参考 input_ids 逐 token 一致，不看解码后的中文是否相似。一处隐蔽点：`merges.txt` 中以 `#` 开头的行是真实合并规则，仅首行 `#version` 是注释；按常规跳过 `#` 开头会丢失部分规则。

### 5. 多模态 embedding 融合

文本 token 先经 embedding 查表得到 `[seq_len, 2048]`，再把 image 占位行按位置精确替换为 vision projector 输出，并校验 image token 数量与 projector 输出行数严格相等。若把占位 id 当普通词表 id 查表，模型将完全看不到图片，却仍会输出一段流畅、看似合理的描述——此类缺陷不崩溃、只表现为"效果不佳"，故必须显式校验。

### 6. LLM 拆 prefill 与 decode 两类图

语言模型 28 层，每层导出两类 ncnn 图：

- **prefill 图**（seq_len=512）：并行处理完整图文 prompt，输出当前层 hidden 与 K/V；
- **decode 图**：只处理最新 token，配合完整 KV cache 与掩码输出 hidden。

decode 进一步拆为两张子图：`dqkv` 图输出 q、k（已施加 RoPE）、v，由 runtime 把当前 token 的 k/v 写入 cache 槽位；`dattn` 图消费完整 cache 与掩码输出 hidden。RoPE、掩码、cache 的生命周期全部由 C++ runtime 显式管理，不留在图内。

embedding 与 lm_head 不进图。模型 `tie_word_embeddings=true`，二者共享一张 `[vocab, 2048]` fp16 词表，以原始二进制存储；runtime 直接查表（embedding）、以 final_norm 后逐行点积取 argmax（lm_head）。此举省去约 1.2GB 的词表大图，且 greedy 解码本就只需 argmax。

### 7. 并行 prefill 与定长窗口 decode

prompt 右填充至 512 后逐层并行 prefill，一次处理 512 个位置，远快于逐 token 串行。本实现的 KV cache 采用**定长窗口**而非动态裁剪：cache 恒为 `[8, 512, 128]`，以加性掩码标记有效前缀（位置 ≤ t 处为 0，其余为 -1e9）。全部张量形状静态，对 ncnn 最友好；prompt padding 由掩码自然屏蔽，无需显式裁剪 cache；当前 token 的 K/V 由 runtime 直接写入 cache 槽位，图内不出现小张量拼接（ncnn 对小张量 Concat/Reshape 是崩溃高发操作）。

### 8. 逐 token decode 与解码

prefill 得到首 logits 后 runtime 取 greedy argmax，之后每轮只处理新 token：embedding → 28 层（dqkv 写 cache、dattn 读 cache）→ final RMSNorm → lm_head argmax。达到 `--max-new-tokens` 或遇到 EOS（151645/151643）时停止，最后由 tokenizer 把 token ids 解码回 UTF-8。

### 9. 构建与运行

CMake 支持 Linux 与 Windows。MinGW-w64 GCC 与 Linux 同源，一次通过；MSVC 需启用 UTF-8 编译选项并定义 `NOMINMAX`。Windows 下中文 prompt 经 argv 传入时为系统区域编码（GBK），须经 `GetCommandLineW` 取 UTF-16 转 UTF-8，并 `SetConsoleOutputCP(CP_UTF8)`，否则 token 数异常、输出跑偏。WSL 下编译 ncnn 时构建目录须置于 ext4（如 `~/`）而非 NTFS，否则 CMake `configure_file` 报权限错误。

## 实现难点

以下为移植中最具代表性的若干问题，每一项均已复现并修复。

### 1. pnnx 无法稳定转换完整多模态模型

原始模型包含 Hugging Face remote code、动态 shape、FlashAttention/SDPA、动态 vision RoPE 与 KV cache，直接整体转换会生成 pnnx 无法稳定解析或 ncnn 无法正确执行的图。解决办法是按计算边界拆分：vision embedding、vision layer、vision norm、projector、LLM layer、text norm、lm_head 分别转换，patch 卷积改写为数学等价的 Linear，动态部分全部上提到 runtime。拆图的另一收益是每张子图可独立与参考比对，使"整链错误"能被二分定位到具体层与算子。

此外，不应直接 trace HF 层——其内部的 dtype 分支与控制流会使 pnnx 解析失败。本实现改用纯 `nn.Linear` 与自写 RMSNorm 重建数学等价的层，从 HF 逐名拷贝权重，并以随机输入对 HF 原层自检（相对误差 < 1e-4）后放行。该自检不可省略：Qwen3 的 q_norm/k_norm 作用于 view 成多头之后的每个 head，手工重建极易在此类细节出错。

### 2. pnnx 的自动算子融合与 HF 语义失配

pnnx 会把若干模式自动融合为 ncnn 内置算子，而融合产物的语义并不总与 HF 一致，且无关闭开关。本实现的对策是把源码改写为数学等价、但融合模式匹配不上的形式：

- **RoPE**：标准 rotate-half 写法会被融合成 `RotaryEmbed`，其轴语义与视觉二维双轴 RoPE 失配。改写为与常量 ±1 置换矩阵 R 的矩阵乘 `q*cos + (q@R)*sin`（R 满足 `(x@R)[i]=-x[i+64] (i<64) / x[i-64] (i>=64)`），pnnx 只看到一次普通 matmul，无从融合。该错误在对称一维位置表下不可观测，只有用真实二维双轴表才暴露——调试探针必须用真实数据。
- **注意力**：`scale+mask+softmax+matmul` 会被融合成 `SDPA`，语义失配且部分平台 kernel 崩溃。改为手工展开 softmax（`amax→sub→exp→sum→div`）。
- **GQA 展开**：`repeat_interleave` 会被 pnnx 静默丢弃（不报错、图照出、结果错，唯一信号是该层 FLOPS 为 0）。改用 HF `repeat_kv` 的 expand+reshape 写法。
- **批轴**：无 batch 维的三维张量会使 pnnx 批轴推断混乱，产生大量 "not supported"。全程保留显式 batch=1、按 `[1,H,N,D]` 布局组织注意力。

导出后另需对 param 做两处后处理：重名图层（嵌套子模块产生的同名 `splitncnn`）加后缀去重；为被非 Split 层直接消费的网络输入插入 Split 隔离层，否则 inplace 层直连输入会运行时崩溃。

### 3. ncnn 在小尺寸张量上越界，decode 图无法直接导出

ncnn 的 RMSNorm 与 UnaryOp（square/rsqrt/pow）在 h 小于 SIMD 打包宽度时（实测 h ≤ 2）越界崩溃，而 decode 的查询长度天然为 1。手写展开 RMSNorm 以规避会陷入双向约束：手写所需的 `x*x` 使 pnnx 在 pass_level3 崩溃，改用 UnaryOp(square)/pow 又触发 ncnn 的小尺寸越界。最终方案是保留原生 RMSNorm（大尺寸张量上完全正常、pnnx 转换干净），将 decode 图按查询长度 8 而非 1 导出：runtime 输入 8 份复制的当前 token embedding，仅取输出第 0 行，cache 写入亦取第 0 行。decode 图单 token 计算量极小，8 倍冗余可忽略。此法可一般化为——框架在小张量上有缺陷时，把张量填充回已验证的形状区间，比修改框架或替换算子更稳妥。

### 4. 视觉链路激活值超出 fp16 表示范围

视觉链路激活值实测可达 4×10⁶，而 fp16 上限为 65504。这是模型固有特性（最后一层前存在巨激活），非缺陷。若视觉侧用 fp16，误差逐层累积并在 tail 溢出为 inf。因此视觉链路必须全程 fp32。补充一处 ncnn Python 接口的坑：`net.opt` 取出的是副本，逐属性赋值会丢失，须整体构造 `Option` 后赋回。

### 5. ncnn Mat 的内存借用语义

ncnn 的 `Mat` 由 numpy 或 C++ 缓冲区构造时不拷贝内存、只借引用。写作 `ncnn.Mat(arr.astype(np.float32))` 时，`.astype` 产生的临时数组随即被回收，前向读的是已释放内存。症状极具迷惑性：Linux 上因内存尚未被覆盖而时对时错，产生余弦相似度 0.3~0.7 的半相关数值；Windows 上直接访问违例。排查中一度误判为"某平台 ncnn 库不稳定"。修法是持有引用直至提取完成（`keep = [np.ascontiguousarray(a, dtype=np.float32) for a in inputs]`）；C++ 侧同理，调用方缓冲区的生命周期须覆盖整个 extract。

### 6. 精度域边界导致的伪分叉

C++ 端到端跑通后，128-token 对齐曾在第 62 步分叉，逐层排查显示每层余弦相似度均 ≥ 0.9999，找不到误差源。根因是参考精度与实现精度不一致：参考为 bf16、实现为 fp32，长程 greedy 中某步两候选 token 的 logit 差小于 bf16 精度粒度，精度更高的一方选出与参考不同的 token，导致 KV cache 分叉、整段发散。将参考改为同精度 fp32 后，128/128 逐 id 完全一致。由此得出的原则是：逐 token 对齐必须在同精度下验收，跨精度比对只能考察余弦相似度、不能考察 token id。

## 调试方法

移植中期主要瓶颈由编码转为定位，为此构建三件工具，将每处缺陷的定位时间压到分钟级：其一，子进程二分崩溃定位器——解析 param 得到拓扑序 blob 列表，对每个 blob 起子进程尝试 extract，崩溃只终止子进程，二分锁定首个崩溃算子；其二，逐 blob 对质——PyTorch 侧 dump 某层全部中间量、ncnn 侧提取对应 blob 算相对误差，首个跳升至 1e-3 以上者即为元凶；其三，param 算子静态审查器——导出后即以文本检查残留融合算子、重名图层、直连输入的 inplace 层与危险 UnaryOp，把多数崩溃拦在耗时的端到端测试之前。工作纪律固化为：导出后先静态审查 param，通过才跑验证。

## 已知限制

- **固定 252×252 方形输入（324 视觉 token）**：图片预处理已实现动态尺寸计算，但视觉图与 vision RoPE 按固定 18×18 grid 导出，非方形/非该分辨率输入会形状失配。支持任意分辨率需按可变 grid 重新导出整套视觉图。
- **上下文窗口固定 512 token**：prefill 图与 KV cache 为定长 512 窗口，融合序列（prompt + 324 图片占位）超过 512 时被显式拒绝并给出提示。加长需重导更大序列的 prefill 图并扩大 cache 窗口。

完整代码、导出脚本、逐级验证脚本与从零复现步骤见代码仓库。欢迎指正与讨论。
